File: SecondYear/semestrul4/MPP/PROIECT/CSharpProjectNetworking/Networking/rpcprotocol/ProjectServicesRpcProxy.cs

using System.Collections.Concurrent;
using System.Net.Sockets;
using System.Text.Json;
using System.Text.Json.Serialization;
using log4net;
using Model;
using Services;

namespace Networking.rpcprotocol
{
    public class ProjectServicesRpcProxy : IProjectServices
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(ProjectServicesRpcProxy));

        private readonly string host;
        private readonly int port;
        private TcpClient connection;
        
        private NetworkStream stream;
        private StreamReader input;
        private StreamWriter output;
        
        private JsonSerializerOptions options;

        private BlockingCollection<Response> qresponses;
        private EventWaitHandle _waitHandle;
        
        private IProjectObserver client;

        private volatile bool connected;


        public ProjectServicesRpcProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            options = new JsonSerializerOptions
            {
                Converters = { new JsonStringEnumConverter() },
                PropertyNameCaseInsensitive = true
            };
            qresponses = new BlockingCollection<Response>();
            logger.Info($"RPC Proxy created for server {host}:{port}");
            InitializeConnection();
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                input = new StreamReader(stream);
                output = new StreamWriter(stream);
                _waitHandle = new AutoResetEvent(false);
                connected = true;
                StartReader();
                logger.Info($"Connected to server {host}:{port}");
            }
            catch (Exception e)
            {
                logger.Error("Error connecting to server: " + e.Message, e);
                throw new ProjectException("Cannot connect to server: " + e.Message, e);
            }
        }
        
        private void CloseConnection()
        {
            connected = false;
            try
            {
                input.Close();
                output.Close();
                stream.Close();
                connection.Close();
                _waitHandle.Close();
                client = null;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }
        
        private void StartReader()
        {
            Thread tw =new Thread(Run);
            tw.Start();
        }

        public virtual void Run()
        {
            while(connected)
            {
                try
                {
                    
                    string responseJson = input.ReadLine();
                    if (string.IsNullOrEmpty(responseJson)) 
                        continue; 
                    Response response = JsonSerializer.Deserialize<Response>(responseJson, options);
                    logger.Debug("response received "+response);
                    if (IsUpdate(response))
                    {
                        HandleUpdate(response);
                    }
                    else
                    {
                        lock (qresponses)
                        {
                            qresponses.Add(response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (Exception e)
                {
                    logger.Error("Reading error "+e);
                }
				
            }
        }
        private bool IsUpdate(Response response)
        {
            return response.Type == ResponseType.UPDATE_INSCRIERE;
        }
        private void HandleUpdate(Response response)
        {
            try
            {
                logger.Info("Handling update response");
                if (IsUpdate(response))
                {
                    Inscriere inscriere = DeserializeResponse<Inscriere>(response.Data, "HandleUpdate");
                    client.InscriereAdded(inscriere);
                    logger.Info("Client updated with inscription info");
                    
                }
            }
            catch (Exception e)
            {
                logger.Error("Error handling update", e);
            }
        }

        private void sendRequest(Request request)
        {
            try
            {
                string jsonRequest = JsonSerializer.Serialize(request, options);
                output.WriteLine(jsonRequest);
                output.Flush();
                
            }
            catch (Exception e)
            {
                throw new ProjectException("Error handling request", e);
            }
        }

        private Response readResponse()
        {
            Response response = null;
            try
            {
                _waitHandle.WaitOne();
                response = qresponses.Take();
            }catch (Exception e) {
                Console.WriteLine(e.StackTrace);
            }
            return response;
        }
        
        private T DeserializeResponse<T>(object responseData, string methodName) where T : class, new()
        {
            try
            {
                if (responseData is JsonElement jsonElement)
                {
                    var options = new JsonSerializerOptions
                    {
                        PropertyNameCaseInsensitive = true
                    };
            
                    return JsonSerializer.Deserialize<T>(jsonElement.GetRawText(), options) ?? new T();
                }
                else if (responseData is T result)
                {
                    return result;
                }
                else
                {
                    if (typeof(T) == typeof(string) && responseData != null)
                    {
                        return responseData.ToString() as T;
                    }
            
                    throw new ProjectException($"Tip de date neașteptat: {responseData.GetType().Name}");
                }
            }
            catch (JsonException jsonEx)
            {
                logger.Error($"[{methodName}] Eroare de deserializare: {jsonEx.Message}", jsonEx);
                Console.Out.WriteLine($"[{methodName}] Eroare de deserializare: {jsonEx.Message}");
                return new T();
            }
        }
        
        public virtual Organizator Login(string username, string password, IProjectObserver observer)
        {
            InitializeConnection();
            Organizator org = new Organizator(username, password);
            Request request = new Request { Type = RequestType.LOGIN, Data = org };
            sendRequest(request);
            Response response = readResponse();
            this.client = observer;
            if (response.Type == ResponseType.ERROR)
            {
                CloseConnection();
                throw new ProjectException("Eroare in Login");
            }

            return DeserializeResponse<Organizator>(response.Data, "Login");
        }

        public virtual void Logout(Organizator user, IProjectObserver observer)
        {
            this.client = observer;
            Request request = new Request { Type = RequestType.LOGOUT, Data = null };
            sendRequest(request);
            Response response = readResponse();
            CloseConnection();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in LOGOUT");
            }
        }
        
        
        public virtual IEnumerable<Proba> GetAllProba()
        {
            Request request = new Request { Type = RequestType.GET_PROBE, Data = null };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in GetAllProba");
            }

            return DeserializeResponse<List<Proba>>(response.Data, "GetAllProba" );
        }

        public IEnumerable<Organizator> GetAllOrganizatori()
        {
            Request request = new Request { Type = RequestType.GET_ORGANIZATORI, Data = null };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in GetAllOrganizatori");
            }
            
            
            return DeserializeResponse<List<Organizator>>(response.Data, "GetAllOrganizatori");
        }


        public IList<ProbaDTO> GetAllProbaDto()
        {
            Request request = new Request { Type = RequestType.GET_PROBE_DTO, Data = null };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                string errorMessage = response.Data?.ToString() ?? "Unknown error";
                throw new ProjectException(errorMessage);
            }
           
            return DeserializeResponse<List<ProbaDTO>>(response.Data, "GetAllOrganizatori");

        }

        public Proba GetProbaByName(string name)
        {
            Request request = new Request { Type = RequestType.GET_PROBA_BY_NAME, Data = name };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in GetProbeByName");
            }

            return DeserializeResponse<Proba>(response.Data, "GetAllOrganizatori");

        }

        public Proba GetProbaByNameAndRange(string name, int range)
        {
            System.Object[] data = new System.Object[2];
            data[0] = name;
            data[1] = range;
            
            Request request = new Request { Type = RequestType.GET_PROBA_BY_NAME_AND_RANGE, Data = data };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in GetProbeByNameandRange");
            }

            return DeserializeResponse<Proba>(response.Data, "GetAllOrganizatori");

        }

        public CategorieVarsta GetVarstaByRange(string selectedCategory)
        {
            Request request = new Request { Type = RequestType.GET_VARSTA_BY_RANGE, Data = selectedCategory };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in GetVarstaByRange");
            }

            return DeserializeResponse<CategorieVarsta>(response.Data, "GetAllOrganizatori");

        }

        public IList<ParticipantDTO> SearchParticipants(Proba selectedProba, CategorieVarsta selectedCategory)
        {
            Object[] data = new System.Object[2];
            data[0] = selectedProba;
            data[1] = selectedCategory;
            
            Request request = new Request { Type = RequestType.SEARCH_PARTICIPANTS, Data = data };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Error in SearchParticipants");
            }

            return DeserializeResponse<List<ParticipantDTO>>(response.Data, "GetAllOrganizatori");
            
        }
        
        private string DeserializeResponseToString(object responseData, string methodName)
        {
            if (responseData is JsonElement jsonElement)
            {
                return jsonElement.GetString() ?? string.Empty;
            }

            if (responseData is string str)
            {
                return str;
            }

            throw new ProjectException($"[{methodName}] Nu se poate converti obiectul în string. Tip primit: {responseData.GetType().Name}");
        }


        public string RegisterParticipant(string name, string cnp, string event1, string event2, string range)
        {
            System.Object[] data = new System.Object[6];
            data[0] = name;
            data[1] = cnp;
            data[2] = event1;
            data[3] = event2;
            data[4] = range;
            
            Request request = new Request { Type = RequestType.REGISTER, Data = data };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in RegisterParticipant");
            }

            return DeserializeResponseToString(response.Data, "RegisterParticipant");
        }

        public Participant? FindParticipantByCNP(string cnp)
        {
            Request request = new Request { Type = RequestType.FIND_PARTICIPANT_BY_CNP, Data = cnp };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in FindParticipantByCNP");
            }

            return DeserializeResponse<Participant>(response.Data, "GetAllOrganizatori");

        }

        public void RefreshProbeStatistics()
        {
            Request request = new Request { Type = RequestType.REFRESH_PROBE_STATISTICS, Data = null };
            sendRequest(request);
            Response response = readResponse();

            if (response.Type == ResponseType.ERROR)
            {
                throw new ProjectException("Eroare in RefreshProbeStatistics");
            }
        }
    }
}

File: SecondYear/semestrul4/MPP/PROIECT/JavaProjectORM/ProjectNetworking/src/main/java/ro/mpp2025/javaprojectui/rpcprotocol/ProjectServicesRpcProxy.java
package ro.mpp2025.javaprojectui.rpcprotocol;


import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2025.javaprojectui.*;
import ro.mpp2025.javaprojectui.dto.ParticipantDTO;
import ro.mpp2025.javaprojectui.dto.ProbaDTO;
import ro.mpp2025.javaprojectui.Participant;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ProjectServicesRpcProxy implements ProjectServices {
    private final String host;
    private final int port;

    private ProjectObserver client;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;

    private final BlockingQueue<Response> qresponses;
    private volatile boolean finished;

    private static final Logger logger = LogManager.getLogger(ProjectServicesRpcProxy.class);

    public ProjectServicesRpcProxy(String host, int port) throws ProjectException {
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<>();
        logger.info("RPC Proxy created for server {}:{}", host, port);
        initializeConnection();
    }

    private void initializeConnection() throws ProjectException {
        try {
            logger.debug("Initializing connection to {}:{}", host, port);
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
            logger.info("Connection established successfully to {}:{}", host, port);
        } catch (IOException e) {
            logger.error("Error initializing connection to {}:{}: {}", host, port, e.getMessage(), e);
            throw new ProjectException("Error connecting to server: " + e.getMessage(), e);
        }
    }


    private void startReader() {
        Thread readerThread = new Thread(() -> {
            logger.info("Starting response reader thread");
            while (!finished) {
                try {
                    Response response = (Response) input.readObject();
                    if (response.getType() == ResponseType.UPDATE_INSCRIERE) {
                        handleUpdateResponse(response);
                    } else {
                        qresponses.put(response);
                    }
                } catch (IOException e) {
                    if (!finished) {
                        logger.error("Error reading response: {}", e.getMessage(), e);
                        closeConnection();
                        break;
                    }
                } catch (ClassNotFoundException ignored) {

                } catch (InterruptedException e) {
                    logger.error("Interrupted while adding response to queue: {}", e.getMessage(), e);
                    Thread.currentThread().interrupt();
                }
            }
        });
        readerThread.setDaemon(true);
        readerThread.start();
    }

    private void handleUpdateResponse(Response response) {
        try {
            if (response.getType() == ResponseType.UPDATE_INSCRIERE) {
                logger.info("Received update notification");
                if (client != null) {
                    Inscriere inscriere = (Inscriere) response.getData();
                    client.inscriereAdded(inscriere);
                } else {
                    logger.warn("Client observer not set, couldn't forward notification");
                }
            }
        } catch (Exception e) {
            logger.error("Error handling update notification: {}", e.getMessage(), e);
        }
    }

    private void closeConnection() {
        logger.info("Closing connection to {}:{}", host, port);
        finished = true;
        try {
            if (input != null) input.close();
        } catch (IOException e) {
            logger.error("Error closing input stream: {}", e.getMessage());
        }
        try {
            if (output != null) output.close();
        } catch (IOException e) {
            logger.error("Error closing output stream: {}", e.getMessage());
        }
        try {
            if (connection != null && !connection.isClosed()) connection.close();
            logger.debug("Socket connection closed.");
        } catch (IOException e) {
            logger.error("Error closing socket connection: {}", e.getMessage());
        }
        this.client = null;
    }


    private Response sendRequestAndWait(Request request) throws ProjectException {
        try {
            logger.debug("Sending request: {}", request);
            synchronized (output){
                output.writeObject(request);
            }
            output.flush();
            Response response = qresponses.take();
            logger.debug("Received response from queue: {}", response);
            return response;
        } catch (IOException e) {
            logger.error("Error sending request or flushing output: {}", e.getMessage(), e);
            closeConnection();
            throw new ProjectException("Error sending request to server: " + e.getMessage(), e);
        } catch (InterruptedException e) {
            logger.error("Interrupted while waiting for response: {}", e.getMessage(), e);
            Thread.currentThread().interrupt();
            closeConnection();
            throw new ProjectException("Interrupted while waiting for server response: " + e.getMessage(), e);
        }
    }


    @Override
    public Organizator login(String username, String password, ProjectObserver client) throws Exception {
        if (this.output == null) {
            initializeConnection();  // Ensure connection is initialized before using it
        }
        Organizator orgToSend = new Organizator(username, password);
        Request req = new Request.Builder().type(RequestType.LOGIN).data(orgToSend).build();

        Response response = sendRequestAndWait(req);

        if (response.getType() == ResponseType.OK) {
            logger.info("Login successful via proxy for user: {}", username);
            this.client = client;
            return (Organizator) response.getData();
        } else {
            logger.warn("Login failed via proxy for user {}: {}", username, response.getData());
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
    }

    @Override
    public void logout(Organizator user, ProjectObserver client) throws ProjectException {
        this.client = client;
        Request req = new Request.Builder().type(RequestType.LOGOUT).data(user).build();

        Response response = sendRequestAndWait(req);
        closeConnection();

        if (response.getType() == ResponseType.ERROR) {
            logger.warn("Logout failed via proxy for user {}: {}", user.getUsername(), response.getData());
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
    }

    @Override
    public Iterable<Organizator> getAllOrganizatori() throws ProjectException {
        Request req = new Request.Builder().type(RequestType.GET_ORGANIZATORI).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }

        return (List<Organizator>) response.getData();
    }

    @Override
    public Iterable<Proba> getAllProba() throws ProjectException {
        Request req = new Request.Builder().type(RequestType.GET_PROBE).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }

        return (List<Proba>) response.getData();
    }

    @Override
    public List<ProbaDTO> getAllProbaDTO() throws ProjectException {
        Request req = new Request.Builder().type(RequestType.GET_PROBE_DTO).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }

        return (List<ProbaDTO>) response.getData();
    }



    @Override
    public Proba getProbaByName(String name) throws ProjectException {
        Request req = new Request.Builder().type(RequestType.GET_PROBA_BY_NAME).data(name).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
        return (Proba) response.getData();
    }


    @Override
    public Proba getProbaByNameAndRange(String name, Integer range) throws ProjectException {
        Object[] data = new Object[2];
        data[0] = name;
        data[1] = range;
        Request req = new Request.Builder().type(RequestType.GET_PROBA_BY_NAME_AND_RANGE).data(data).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
        return (Proba) response.getData();
    }

    @Override
    public CategorieVarsta getVarstaByRange(String selectedCategory) throws ProjectException {
        Request req = new Request.Builder().type(RequestType.GET_VARSTA_BY_RANGE).data(selectedCategory).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
        return (CategorieVarsta) response.getData();
    }

    @Override
    public List<ParticipantDTO> searchParticipants(Proba selectedProba, CategorieVarsta selectedCategory) throws ProjectException {
        Object[] searchData = new Object[2];
        searchData[0] = selectedProba;
        searchData[1] = selectedCategory;
        Request req = new Request.Builder().type(RequestType.SEARCH_PARTICIPANTS).data(searchData).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
        return (List<ParticipantDTO>) response.getData();
    }

    @Override
    public String registerParticipant(String name, String cnp, String event1, String event2, String range) throws ProjectException {
        Object[] registrationData = new Object[5];
        registrationData[0] = name;
        registrationData[1] = cnp;
        registrationData[2] = event1;
        registrationData[3] = event2;
        registrationData[4] = range;
        Request req = new Request.Builder().type(RequestType.REGISTER).data(registrationData).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
        return (String) response.getData();
    }

    @Override
    public Optional<Participant> findParticipantByCNP(String cnp) throws ProjectException {
        Request req = new Request.Builder().type(RequestType.FIND_PARTICIPANT_BY_CNP).data(cnp).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown login error";
            throw new ProjectException(errorMessage);
        }
        return (Optional<Participant>) response.getData();
    }

    @Override
    public void refreshProbeStatistics() throws ProjectException {
        Request req = new Request.Builder().type(RequestType.REFRESH_PROBE_STATISTICS).build();
        Response response = sendRequestAndWait(req);
        if (response.getType() == ResponseType.ERROR) {
            String errorMessage = (response.getData() != null) ? response.getData().toString() : "Unknown error";
            throw new ProjectException(errorMessage);
        }
    }
}
